interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,392 @@
+//
+// If the prompt was merged with an identical existing prompt, returns the
+// existing prompt and true, indicating it was merged. If a new prompt was
+// added, returns the new prompt and false, indicating the prompt was not
+// merged.
+//
+// The caller must ensure that the given permissions are in the order in which
+// they appear in the available permissions list for the given interface.
+func (pdb *PromptDB) AddOrMerge(metadata *prompting.Metadata, path string, permissions []string, listenerReq *listener.Request) (*Prompt, bool, error) {
+	availablePermissions, err := prompting.AvailablePermissions(metadata.Interface)
+	if err != nil {
+		// Error should be impossible, since caller has already validated that
+		// iface is valid, and tests check that all valid interfaces have valid
+		// available permissions returned by AvailablePermissions.
+		return nil, false, err
+	}
+
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+
+	userEntry, ok := pdb.perUser[metadata.User]
+	if !ok {
+		pdb.perUser[metadata.User] = &userPromptDB{
+			ByID: make(map[string]*Prompt),
+		}
+		userEntry = pdb.perUser[metadata.User]
+	}
+
+	constraints := &PromptConstraints{
+		Path:                 path,
+		Permissions:          permissions,
+		AvailablePermissions: availablePermissions,
+	}
+
+	// Search for an identical existing prompt, merge if found
+	for _, prompt := range userEntry.ByID {
+		if prompt.Snap == metadata.Snap && prompt.Interface == metadata.Interface && prompt.Constraints.equals(constraints) {
+			prompt.listenerReqs = append(prompt.listenerReqs, listenerReq)
+			// Although the prompt itself has not changed, re-record a notice
+			// to re-notify clients to respond to this request. A client may
+			// have replied with a malformed response and not retried after
+			// receiving the error, so this notice encourages it to try again
+			// if the user retries the operation.
+			pdb.notifyPrompt(metadata.User, prompt.ID, nil)
+			return prompt, true, nil
+		}
+	}
+
+	if len(userEntry.ByID) >= maxOutstandingPromptsPerUser {
+		logger.Noticef("WARNING: too many outstanding prompts for user %d; auto-denying new one", metadata.User)
+		sendReply(listenerReq, false)
+		return nil, false, ErrTooManyPrompts
+	}
+
+	id := pdb.nextID()
+	timestamp := time.Now()
+	prompt := &Prompt{
+		ID:           id,
+		Timestamp:    timestamp,
+		Snap:         metadata.Snap,
+		Interface:    metadata.Interface,
+		Constraints:  constraints,
+		listenerReqs: []*listener.Request{listenerReq},
+	}
+	userEntry.ByID[id] = prompt
+	pdb.notifyPrompt(metadata.User, id, nil)
+	return prompt, false, nil
+}
+
+// Prompts returns a slice of all outstanding prompts.
+func (pdb *PromptDB) Prompts(user uint32) []*Prompt {
+	pdb.mutex.RLock()
+	defer pdb.mutex.RUnlock()
+	userEntry, ok := pdb.perUser[user]
+	if !ok || len(userEntry.ByID) == 0 {
+		return nil
+	}
+	prompts := make([]*Prompt, 0, len(userEntry.ByID))
+	for _, prompt := range userEntry.ByID {
```

Comment:
If we tweak the way userEntry.ByID is stored this could return the slice directly or copy the slice and return it.